### PR TITLE
Digest-generic MerkleTreeGadget

### PR DESCRIPTION
**Pull request by tessico**
## Description

Making the circuit equivalent of MerkleTree generic over a digest algorithm is more tricky, because both the `MerkleTreeGadget` and `DigestAlgorithmGadget` need to be implemented for `PlonkCircuit`.

I'm proposing two designs here. One is to add it as a generic parameter to `MerkleTreeGadget`, see [the last commit](https://github.com/EspressoSystems/jellyfish/pull/167/commits/4e3a09c80fac905b26f2632f1d4e33ece9bd3398). The other option is to add `DigestAlgorithmGadget` as an associated type on the `MerkleTreeGadget` trait, see [here](https://github.com/EspressoSystems/jellyfish/pull/167/commits/371e87470b2261ee1ea459d9f7e87a5d634d0397). Both options work, it's just a matter of convenience and ergonomics. Please let me know your opinions.

(Note that initially I thought that the associated type would not work in the case where in the future we need to provide multiple implementations of the `MerkleTreeGadget` trait. But since the trait already is generic over `MerkleTreeScheme`, which, implicitly, will be using the counterpart of `DigestAlgorithmGadget`, it should still be possible to provide multiple impl even with `DigestAlgorithmGadget` being an associated type. See [the rust book](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#specifying-placeholder-types-in-trait-definitions-with-associated-types) for more explanation on the difference between the two.)

closes: #142 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (main)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (mrain) [inline primitives/src/circuit/merkle_tree/mod.rs:42]: Why does it have to be `?Sized`?

**Review by mrain — APPROVED** [review]
Generally LGTM.

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (alxiong) [reply primitives/src/circuit/merkle_tree/mod.rs:42]: Usually it's because we use it in dynamic trait objects somewhere

File: primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs
Comment (alxiong) [inline primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:45]: I'm actually little unsure.
Ideally, shouldn't `DigestAlgorithmGadget` only implemented for `MerkleTreeGadget`? We don't need these `digest_leaf` gadget exposing to the general `PlonkCircuit` IMO.

this may solve our `dyn DigestAlgorithmGadget` issue at the same time.

Comment (tessico) [reply primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:45]: @alxiong Since `MerkleTreeGadget` is a trait, that becomes a little problematic. You either need to use `dyn MerkleTreeGadget` here instead, or do something like:
```
impl<T> DigestAlgorithmGadget for T
where
    T: MerkleTreeGadget
```

which I'm not even sure is possible, since we need to pass in the parameter `F` somehow like `impl<T, F>` (because we want to restrict `T: MerkleTreeGadget<RescueMerkleTree<F>>`), but the trait `DigestAlgorithmGadget` doesn't itself restrict `F`, and we'd run into errors.
By all means please drop a few lines/PR if you have an alternative of how to implement `DigestAlgorithmGadget` only for `MerkleTreeGadget` and not entire `PlonkCircuit` in this design.

But before you do that, please first checkout https://github.com/EspressoSystems/jellyfish/pull/167/commits/371e87470b2261ee1ea459d9f7e87a5d634d0397, which has an alternative design, and actually avoids the issue you mention altogether. (note that the concrete location of the structs/impls there is up for discussion, I might move them later to where `RescueNativeGadget` is defined).

Comment (alxiong) [reply primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:45]: I kind of like your alternative solution, particularly we don't have the `dyn` issue, also compliant with my intuition that `DigestGadget` should be associated/part of/used under the context of `MerkleTreeGadget`.

**Comment by tessico** [comment]
As per [pdiscussion](https://github.com/EspressoSystems/jellyfish/commit/371e87470b2261ee1ea459d9f7e87a5d634d0397) on the first commit, I've gone back to the initial solution by reverting the second commit.

File: primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs
Comment (alxiong) [inline primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:187]: this is subtyping relationship is slightly strange. 

What do you think of `MerkleTreeHelperGadget<F: RescueParameter, D: DigestAlgorithmGadget<F>>`?

Comment (tessico) [reply primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:187]: Upon review against [the other PR](https://github.com/EspressoSystems/jellyfish/pull/172/), I've decided to wait until that non-membership proof PR is merged, as the trait definitions will change there anyway. This PR is smaller and can be adapted more easily.

Comment (tessico) [reply primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:187]: I've added a `blocked` label for clarification. Don't wanna duplicate the work to refactor traits here twice.

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (mrain) [inline primitives/src/circuit/merkle_tree/mod.rs:277]: Should this `pos` be of type `Variable`? Although that's `usize`.

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (mrain) [reply primitives/src/circuit/merkle_tree/mod.rs:63]: I tried the following and it's working:

```
pub trait MerkleTreeGadget<M>
where
    M: MerkleTreeScheme,
    M::NodeValue: PrimeField,
{
    ...
    /// Gadget for the digest algorithm.
    type DigestGadget: DigestAlgorithmGadget<M::NodeValue>;
```

**Review by mrain — CHANGES_REQUESTED** [review]
Excellent work! Will approve if comments are addressed.

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (tessico) [reply primitives/src/circuit/merkle_tree/mod.rs]: Great suggestion @mrain! Done in https://github.com/EspressoSystems/jellyfish/pull/167/commits/abc72a7730547d893c7071678b67f581a286dd05

File: primitives/src/circuit/merkle_tree/mod.rs
Comment (tessico) [reply primitives/src/circuit/merkle_tree/mod.rs]: Good catch. Done in https://github.com/EspressoSystems/jellyfish/pull/167/commits/d3560a6b9cf692989587fc43f39f2dbe4241ddf1

**Review by mrain — APPROVED** [review]
LGTM

**Pull request merged**
